### DISCUSSION TITLE
大家能讲讲自己怎么分流吗

### DISCUSSION THREAD

**Discussion by kls1991** [IrineSistiana/mosdns#630] [Q&A]
小弟才疏学浅（懂一点点编程和linux，网络基本不懂，肯学），理解错误的地方还请见谅，盼大家不吝赐教

----------
我现在理解，多数人是搭建v2ray/clash之类的socks5代理， 由socks5代理接管所有流量，再按规则分流
这样算完美吗？
我认为代理是重要的，但没必要扮演这么重要的角色吧？我梦想把网络环境打造成：
1. 国内网站尽可能少的受影响
2. GitHub之类只是dns污染的站点，即便远程节点甚至代理程序挂了也能访问
3. 代理正常的时候，能访问一切

我现在最需要指点迷津的是：
1. 上面设想，是否应该通过linux 的ipset + nftables 实现？如果是的话，mosdns的用户们是否更偏爱这种分流方式？
2. Windows的路由表能否导入geoip/getsite，黑名单进tun2socks的虚拟网卡，虚拟网卡无脑代理
3. 一次网络的请求，走mosdns的时候就已经可以判断走不走代理了吧？还一定要被“我设想中那个负责分流的家伙”再判断一次吗？再被geoip/geosite筛选一次的话，总感觉有点丑陋，mosdns的mark能否用在这里？

---------------------------------------------
非常想知道大家的分流的设计，很迷茫，求指路……

**Comment by 2cydg** [comment]
主路由mosdns+旁路由clash(fake-ip)分流

**Reply by lingkai995** [reply]
求问大佬，如何让mosdns返回fake ip，我这里remote上游设置了surge的dns198.18.0.2。如果用tcp向mosdns查询的话，能返回surge的fake ip段198.18.0.0/16内的ip；但是用udp向mosdns查询，则会返回域名对应的真实ip。

**Reply by luweijun1992** [reply]
> 主路由mosdns+旁路由clash(fake-ip)分流

我当前是终端的dns是pi-hole、然后pi-hole转发到mosdns进行分流，国内域名直接几个公共dns，国外域名使用OpenClash，分别进行解析。
当前遇到打开网页时好时不好。
我感觉是dns缓存问题，比如pi-hole的解析记录还有缓存，且跟当前OpenClash解析的IP不一致，造成流量到达OpenClash后不是一致导致的。也不一定是这个原因。

**Comment by wslpdgg** [comment]
我的理解，mosdns直接对域名或ip进行判断，国内的流量直接走本地出去，国外的无脑走科学渠道

**Reply by kinist** [reply]
代理用什么？clash吗？

**Comment by nil-ca** [comment]
mosdns的ipset和nftset好像就可以，但是目前还不会用。
要是用FAKE-IP
如果是单设备同时用mosdns和clash 需要自己学一学iptables或者nftables才能分流，按照shellclash openclash的配置来看，都是把所有流量都接管了的，记得好像是在面板里面就能看到国内的IP都在里面过一遍。
如果是旁路由就没这么麻烦，直接在路由里面加一条到198.18.0.0/16到clash设备的路由就好了。

**Reply by kinist** [reply]
为啥旁路由的话，只需要在主路由加条规则呢？不是太明白。

**Reply by nil-ca** [reply]
其实都是加规则
旁路由的话，只需要告诉客户端你要访问的域名是198.18的IP去找clash的旁路由就行，像软路由都有加路由的页面，很简单。
单设备的话要用iptables或者nftables告诉客户端你要访问的域名是198的ip，要去某个端口，添加方式稍微比旁路由绕一点。

**Reply by luweijun1992** [reply]
> 为啥旁路由的话，只需要在主路由加条规则呢？不是太明白。

因为OpenClash的Fake-IP模式所有域名解析都是1928.18.0.0/18范围内得IP，所有只需要在主路由上写一条静态路由198.18.0.0/16指向OpenClash就可以了。

**Reply by kinist** [reply]
局域网设备DNS指向OpenClash，解析后的地址不都是198的吗，不管是国内或国外。局域网设备DNS到底指向mosdns还是clash？

**Reply by nil-ca** [reply]
都可以，有两种用法，可以是  设备->clash->mosdns ，也可以是  设备->mosdns->clash。
第一种则是全部流量过一遍clash后再请求mosdns，相当于二次分流，可以让mosdns帮clash去广告和加DNS缓存。虽然DNS性能差别不大，但是对于小设备来说国内IP也过一遍clash会导致CPU不够吃。
第二种是用规则让国内IP不经过clash，路由规则要自己做，因为DNS解析后变成198.18后没有路由，所以需要自己做一条，不吃clash的CPU。

可以看一下这个 #470 

www.v2ex.com/t/892373
这里引用一下的十楼，他说的就是这两种模式。
> Clash 如果用官方推荐的“Fake-ip”模式，那所有的连接都会从 Clash 内核发起（正常来说路由器是为“转发”优化的，发起连接是终端干的活），这样域名解析可以加速一些，能提升访问网页的体验，没有 dns 污染的问题，但对代理设备的要求较高，这种无线设备来处理很勉强。也有坏处，比如影响一些游戏之类，个人觉得这模式挺蠢的。。。
推荐使用“Redir-host”模式，再在 Clash 上游套一个 mosdns 处理域名查询请求，开启 lazy_cache ，这样域名查询速度几乎与“Fake-ip”相当，而且能拿到真正的、无污染的域名 IP 。也不用担心 cdn 优化的问题，因为 Clash 拿到 IP 只是用匹配规则，代理的时候会替换成 fqdn ，再在服务端重新解析域名。
另外，可以把中国 IP 段用 iptables/ipset 绕过 Clash ，直接直连，Clash 仅处理真正需要代理的连接，其发起连接的压力会小许多。

**Comment by anonsaber** [comment]
> 小弟才疏学浅（懂一点点编程和linux，网络基本不懂，肯学），理解错误的地方还请见谅，盼大家不吝赐教
> 
> 我现在理解，多数人是搭建v2ray/clash之类的socks5代理， 由socks5代理接管所有流量，再按规则分流 这样算完美吗？ 我认为代理是重要的，但没必要扮演这么重要的角色吧？我梦想把网络环境打造成：
> 
>     1. 国内网站尽可能少的受影响
> 
>     2. GitHub之类只是dns污染的站点，即便远程节点甚至代理程序挂了也能访问
> 
>     3. 代理正常的时候，能访问一切
> 
> 
> 我现在最需要指点迷津的是：
> 
>     1. 上面设想，是否应该通过linux 的ipset + nftables 实现？如果是的话，mosdns的用户们是否更偏爱这种分流方式？
> 
>     2. Windows的路由表能否导入geoip/getsite，黑名单进tun2socks的虚拟网卡，虚拟网卡无脑代理
> 
>     3. 一次网络的请求，走mosdns的时候就已经可以判断走不走代理了吧？还一定要被“我设想中那个负责分流的家伙”再判断一次吗？再被geoip/geosite筛选一次的话，总感觉有点丑陋，mosdns的mark能否用在这里？
> 
> 
> 非常想知道大家的分流的设计，很迷茫，求指路……

DNS 方面请参考  https://github.com/anonsaber/Homelab/discussions/2
网络路由方面请参考 https://github.com/anonsaber/Homelab/discussions/4

注意，此方案 DNS 与网络路由之间无任何关联性，网络与路由方面使用通用的网络协议进行处理，完全与 DNS 无关。

DNS 只做 DNS 的事儿，路由分流只做路由分流的事儿。完全孤立。

完全满足需求 1 和 3，国内网址完全不受影响。代理正常的时候，能访问一切。
对于需求 2，DNS 的上游需要保证存活可用，可以考虑多写几个上游。

建议搭配商业路由器使用，开销低，性能好，过往的使用经验中 0 误判流量。
